### PR TITLE
Avoid creating TrainerConfig during parse_conf_file()

### PR DESCRIPTION
**Pull request by emailweixu**
Before this change, get_observation() will create a TrainerConfig(). This makes the configurable
values of TrainerConfig to be all labelled as used. And it will not be allowed to change after get_observation().

Now we use get_config() to get the value of TrainerConfig.data_transformer_ctor
to avoid creating TrainerConfig object. So the configs of other TrainerConfig values
can still be changed after get_observation_spec().

Also changed sac_carla.gin to use alf-config

### CODE REVIEW

File: alf/environments/suite_carla.py [hunk 1]
```diff
@@ -1027,7 +1027,7 @@ def __init__(self,
         Args:
             batch_size (int): the number of learning vehicles.
             map_name (str): the name of the map (e.g. "Town01")
-            vehicle_filter (str): the filter for getting vehicle blueprints.
+            vehicle_filter (str): the filter for getting ego vehicle blueprints.
```
Comment (hnyu) [inline alf/environments/suite_carla.py]: I think 'ego' is a little confusing here. Might just explain that this filter is used for vehicles that will be trained to drive. And the other vehicles will use a catch-all filter.

File: alf/bin/train.py [hunk 2]
```diff
@@ -87,9 +85,17 @@ def train_eval(ml_type, root_dir):
...
+    random_seed = alf.get_config('TrainerConfig.random_seed')
+    alf.config(
+        'TrainerConfig',
+        random_seed=common.set_random_seed(random_seed),
+        raise_if_used=False)
     conf_file = common.get_conf_file()
-    common.parse_conf_file(conf_file)
-    train_eval(FLAGS.ml_type, FLAGS.root_dir)
+    try:
+        common.parse_conf_file(conf_file)
```
Comment (le-horizon) [inline alf/bin/train.py:90]: will it raise error if parse_conf_file tries to set random seed again?

File: alf/bin/train.py [hunk 3]
```diff
@@ -87,9 +85,17 @@ def train_eval(ml_type, root_dir):
 
 def main(_):
     FLAGS.alsologtostderr = True
+    random_seed = alf.get_config('TrainerConfig.random_seed')
```
Comment (le-horizon) [inline alf/bin/train.py]: will this only read from the command line config?  What's the reason for making this change: first get random_seed, then set it and set config, then parse conf file?  It seems complicated and prone to error?

**Review by le-horizon — COMMENTED** [review]
Some values need to be configured before parse_alf_config, so the config and initialization of alf seems intertwined, and a bit hard to follow.  We may need to create some documentation for this, in case we forget in the future?

File: alf/config_util.py [hunk 4]
```diff
@@ -189,20 +194,8 @@ def is_used(self):
...
-        mutable (bool): whether the config can be changed later. If the user
-            tries to change an existing immutable config, the change will be
-            ignored and a warning will be generated. You can always change a
-            mutable config. ``ValueError`` will be raised if trying to set a new
-            immutable value to an existing immutable value.
-
-
-    """
+def _get_config_node(config_name):
+    """Get the _Config corresponding to config_name."""
```
Comment (Haichao-Zhang) [inline alf/config_util.py]: Does ```_Config``` has any special meaning here? Maybe can change to ```config node```.

File: alf/config_util.py [hunk 5]
```diff
@@ -224,7 +217,26 @@ def config1(config_name, value, mutable=True):
...
+            tries to change an existing immutable config, the change will be
+            ignored and a warning will be generated. You can always change a
+            mutable config. ``ValueError`` will be raised if trying to set a new
+            immutable value to an existing immutable value.
+        raise_if_used (bool): If True, ValueError will be raised if trying to
+            config a value which has already been used.
+    """
+    config_node = _get_config_node(config_name)
+
+    if raise_if_used and config_node.is_used():
```
Comment (Haichao-Zhang) [inline alf/config_util.py:240]: When raise_if_used is False,  maybe we can generate a warning message

File: alf/config_util.py [hunk 6]
```diff
@@ -250,6 +262,28 @@ def config1(config_name, value, mutable=True):
         config_node.set_mutable(mutable)
 
 
+def get_config(config_name):
```
Comment (Haichao-Zhang) [inline alf/config_util.py]: Maybe we can change the name of this function to ```get_config_value```?

File: alf/bin/train.py (hunk 2 above)
Comment (emailweixu) [reply alf/bin/train.py:90]: moved all random_seed ralated code to get_env

File: alf/config_util.py (hunk 4 above)
Comment (emailweixu) [reply alf/config_util.py]: _Config is the class name

File: alf/config_util.py (hunk 5 above)
Comment (emailweixu) [reply alf/config_util.py:240]: With the current way using raise_if_used=True inside config_helper, the config is always used, and a warning message will be quite confusing.

File: alf/config_util.py (hunk 6 above)
Comment (emailweixu) [reply alf/config_util.py]: changed

File: alf/environments/suite_carla.py (hunk 1 above)
Comment (emailweixu) [reply alf/environments/suite_carla.py]: Updated.

File: alf/bin/train.py (hunk 3 above)
Comment (emailweixu) [reply alf/bin/train.py]: moved to config_helpers.get_env() and commented there.

File: alf/examples/sac_carla_conf.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+        mini_batch_length=4,
+        data_transformer_ctor=[ImageScaleTransformer, ObservationNormalizer])
+
+alf.config(
+    'suite_carla.Player',
+    # Not yet able to successfully train with sparse reward.
+    sparse_reward=False,
+    # Cannot train well with allow_negative_distance_reward=False yet.
+    allow_negative_distance_reward=True,
+    # Currently, even a small penalty such as one make the training much worse
```
Comment (le-horizon) [inline alf/examples/sac_carla_conf.py]: make->makes

File: alf/examples/ac_cart_pole_conf.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+# reward scaling
+alf.config('TrainerConfig', data_transformer_ctor=RewardScaling)
+alf.config('RewardScaling', scale=0.01)
+
+# algorithm config
+alf.config('ActorDistributionNetwork', fc_layer_params=(100, ))
+alf.config('ValueNetwork', fc_layer_params=(100, ))
+alf.config(
+    'ActorCriticAlgorithm',
+    optimizer=alf.optimizers.Adam(lr=1e-3, gradient_clipping=10.0))
```
Comment (le-horizon) [inline alf/examples/ac_cart_pole_conf.py:33]: It's fine as is.  Does it make sense to have lr configured using a same named variable in TrainerConfig?

File: alf/config_helpers.py [hunk 9]
```diff
@@ -94,19 +99,71 @@ def get_action_spec():
...
     global _env
     if _env is None:
-        trainer_config = TrainerConfig(root_dir='')
-        _env = create_environment(seed=trainer_config.random_seed)
+        if _is_parsing:
+            random_seed = get_config_value('TrainerConfig.random_seed')
+        else:
+            # We construct a TrainerConfig object here so that the value
+            # configured through gin-config can be properly retrieved.
+            train_config = TrainerConfig(root_dir='')
```
Comment (le-horizon) [inline alf/config_helpers.py:159]: I'm confused.  This config_helper.py is only used by alf config, not gin-config?  Why do we need to care about gin-config here?

File: alf/examples/sac_carla_conf.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+    continuous_projection_net_ctor=proj_net)
+
+env = alf.get_env()
+
+critic_network_cls = partial(
+    alf.networks.CriticNetwork,
+    joint_fc_layer_params=fc_layers_params,
+    activation=activation,
+    use_fc_bn=use_batch_normalization,
+    output_tensor_spec=env.reward_spec())
```
Comment (le-horizon) [inline alf/examples/sac_carla_conf.py:115]: maybe have a common.get_reward_spec() function, and hide alf.get_env() inside it?

File: alf/bin/train.py (hunk 2 above)
Comment (le-horizon) [reply alf/bin/train.py:90]: Looks nice!

**Review by le-horizon — COMMENTED** [review]
Accidentally submitted the partial review.
Also took a look at the new commit.

It's much cleaner now, only a few nits/suggestions if they make sense.

Thanks,
Le

File: alf/examples/sac_carla_conf.py (hunk 7 above)
Comment (emailweixu) [reply alf/examples/sac_carla_conf.py]: removed this comment as it is not accurate anymore

File: alf/examples/sac_carla_conf.py (hunk 10 above)
Comment (emailweixu) [reply alf/examples/sac_carla_conf.py:115]: There are many other env member function. It is not a good idea to do this for all of them. The reason of having get_observation_spec() is that it has additional handling of data_transformer.

File: alf/config_helpers.py (hunk 9 above)
Comment (emailweixu) [reply alf/config_helpers.py:159]: Because get_env() will be used by policy_trainer to get the global env even when gin-config is used for configuration.

File: alf/examples/ac_cart_pole_conf.py (hunk 8 above)
Comment (emailweixu) [reply alf/examples/ac_cart_pole_conf.py:33]: There could be multiple optimizers.

**Review by le-horizon — APPROVED** [review]
LG.

**Review by Haichao-Zhang — APPROVED** [review]
